### PR TITLE
Issue #416

### LINKED ISSUE #416: Update Session Generation
**Is your feature request related to a problem? Please describe.**
The legacy code is not up to current specification

**Describe the solution you'd like**
Update the legacy session generation with a focus on readability.

**(Optional) Include a BDD test that would satisfy the expected behaviour of the new feature**
Needs to pass legacy BDD

### PR DESCRIPTION
**Pull request by andrewnguyen22**
Closes #416

### CODE REVIEW

File: tests/bdd/session/legacySession_test.go [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+
+// ************************************************************************************************************
+// Milestone: Sessions
+//
+// Tentative Timeline (4-6 weeks)
+//
+// Unanswered Questions?
+// - Where is the seed data coming from? 99% sure -> *WORLD STATE*
+// - Is service request validation ddos safe?
+// - Forking behavior
```
Comment (luyzdeleon) [inline tests/bdd/session/legacySession_test.go:20]: You can remove this one, as tendermint won't allow forks.

File: x/session/session.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+type Session struct {
+	SessionKey     SessionKey        `json:"sessionkey"`
+	Developer      SessionDeveloper  `json:"developer"`
+	NonNativeChain SessionBlockchain `json:"nonnativechain"`
+	BlockID        SessionBlockID    `json:"latestBlock"`
+	Nodes          SessionNodes      `json:"sessionNodes"`
+}
+
+// Create a new session from seed data
```
Comment (luyzdeleon) [inline x/session/session.go:11]: We should at least comment the parameters?

Comment (luyzdeleon) [reply x/session/session.go:11]: Also `SessionDeveloper` should be renamed to `SessionApplication`.

File: tests/bdd/session/session_test.go [hunk 3]
```diff
@@ -1,264 +1,82 @@
...
-				Context("No nodes are associated with a blockchain", func() {
-					noNodesSeed := legacy.SessionSeed{DevID: devid, BlockHash: blockhash, RequestedChain: legacy.SHA3FromString("foo"), NodeList: nodelist, Capacity: capacity}
-					It("should return `invalid blockchain` error", func() {
-						_, err := legacy.NewSession(noNodesSeed)
-						Expect(err).To(Equal(legacy.InsufficientNodesError))
-					})
+			Context("All seed data is valid", func() {
+				It("should return nil error", func() {
+					_, err := session.NewSessionKey(validDeveloper, validNonNativeChain, validBlockID)
+					gomega.Expect(err).To(gomega.BeNil())
```
Comment (luyzdeleon) [inline tests/bdd/session/session_test.go:44]: Shouldn't this also test that the new session key is not nil?

File: tests/bdd/session/session_test.go [hunk 4]
```diff
@@ -1,264 +1,82 @@
...
+				It("should return nil error", func() {
+					_, err := session.NewSessionNodes(validNonNativeChain, validSessionKey)
+					gomega.Expect(err).To(gomega.BeNil())
 				})
 			})
 		})
+		Context("Valid seed data for the session", func() {
+			It("should return nil error", func() {
+				_, err := session.NewSession(validDeveloper, validNonNativeChain, validBlockID)
+				gomega.Expect(err).To(gomega.BeNil())
```
Comment (luyzdeleon) [inline tests/bdd/session/session_test.go:80]: Shouldn't this also test that the returned session is not nil?

File: tests/bdd/session/session_test.go [hunk 5]
```diff
@@ -1,264 +1,82 @@
...
+			})
+		})
 	})
+	// abide by the session 'rules'
+	PDescribe("Rules", func() {})
+	// gracefully terminate the session
+	PDescribe("Termination", func() {})
```
Comment (luyzdeleon) [inline tests/bdd/session/session_test.go:90]: I don't think Sessions should ever be "terminated".

File: x/session/sessionNodes.go [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+	// return NodeDistance array to be ordered
+	xorResult, err = xor(result, sessionKey)
+	if err != nil {
+		return nil, err
+	}
+	// sort the nodes based off of distance
+	result = revSort(xorResult)
+
+	// return the top 5 nodes
+	return result[0:5], nil
```
Comment (luyzdeleon) [inline x/session/sessionNodes.go]: Won't this error out if there are less than 5 nodes? Also the `5` should be made a constant in case it needs to be referenced from different places.

File: x/session/sessionNodes.go [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+	}
+	if len(result) == 0 {
+		return nil, InsufficientNodesError
+	}
+	return result, nil
+}
+
+// xor the sessionNodes.publicKey against the sessionKey to find the computationally
+// closest nodes
+func xor(sessionNodes SessionNodes, sessionkey SessionKey) (NodeDistances, error) {
```
Comment (luyzdeleon) [inline x/session/sessionNodes.go:86]: Shouldn't it be better for randomness to xor the address? Since it's a hash of the public key?

File: x/session/sessionNodes.go [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+// returns the length of the node pool -> needed for sort.Sort() interface
+func (n NodeDistances) Len() int { return len(n) }
+
+// swaps two elements in the node pool -> needed for sort.Sort() interface
+func (n NodeDistances) Swap(i, j int) { n[i], n[j] = n[j], n[i] }
+
+// returns if node i is less than node j by XOR value
+// it assumes big endian encoding
+func (n NodeDistances) Less(i, j int) bool {
```
Comment (luyzdeleon) [inline x/session/sessionNodes.go:126]: Where is this being used?

File: tests/bdd/session/legacySession_test.go (hunk 1 above)
Comment (andrewnguyen22) [reply tests/bdd/session/legacySession_test.go:20]: Yes! it's in legacy so I left it but do you think I should remove it?

File: tests/bdd/session/session_test.go (hunk 3 above)
Comment (andrewnguyen22) [reply tests/bdd/session/session_test.go:44]: Yes, great point

File: tests/bdd/session/session_test.go (hunk 5 above)
Comment (andrewnguyen22) [reply tests/bdd/session/session_test.go:90]: "termination" is what we called it in the full spec

The DOS signal that ends the 'session'

File: x/session/sessionNodes.go (hunk 6 above)
Comment (andrewnguyen22) [reply x/session/sessionNodes.go]: Yes, and it is a constant! I just didn't use it lol

File: x/session/sessionNodes.go (hunk 7 above)
Comment (andrewnguyen22) [reply x/session/sessionNodes.go:86]: I re-hash the public key to ensure that it's the same length as the sessionKey

File: x/session/sessionNodes.go (hunk 8 above)
Comment (andrewnguyen22) [reply x/session/sessionNodes.go:126]: It's part of the interface needed for sort.Sort

File: x/session/session.go (hunk 2 above)
Comment (andrewnguyen22) [reply x/session/session.go:11]: Did we officially decide on the nomenclature Application? What was the reasoning behind this decision?

File: tests/bdd/session/session_test.go (hunk 4 above)
Comment (andrewnguyen22) [reply tests/bdd/session/session_test.go:80]: Yes!

File: tests/bdd/session/legacySession_test.go (hunk 1 above)
Comment (luyzdeleon) [reply tests/bdd/session/legacySession_test.go:20]: Yeah, you're right, since it's legacy we can keep it for future reference.

File: x/session/session.go (hunk 2 above)
Comment (luyzdeleon) [reply x/session/session.go:11]: The reasoning is that it's more clear to understand the terminology under the following list:

- Applications
- ApplicationClient/User
- Node

`Applications` is what is actually being served by the network.

File: tests/bdd/session/session_test.go (hunk 5 above)
Comment (luyzdeleon) [reply tests/bdd/session/session_test.go:90]: That means expiration not termination. Maybe just change the term to `Expiration`.

File: x/session/sessionNodes.go (hunk 7 above)
Comment (luyzdeleon) [reply x/session/sessionNodes.go:86]: Got it, that makes sense.

File: x/session/sessionNodes.go (hunk 8 above)
Comment (luyzdeleon) [reply x/session/sessionNodes.go:126]: Got it, a comment about that like you did in the 2 methods above would help identify that.

**Review by luyzdeleon — APPROVED** [review]
